### PR TITLE
Feature/yaxlib visual studio debugger friendly

### PR DESCRIPTION
**Pull request by ycherkes**
Hi YAXLib team.

I want to use your fantastic library in my [Object Dumper extension for Visual Studio](https://github.com/ycherkes/ObjectDumper/tree/experimental-use-yaxlib-instead-of-jsonnet).

[Experimental version of extension](https://github.com/ycherkes/ObjectDumper/raw/experimental-use-yaxlib-instead-of-jsonnet/ObjectDumper.vsix)

**What is done in this PR**:

**Deprecation**:
* [IMemberContext:](https://github.com/YAXLib/YAXLib/blob/612f849d5a0ff47c617c0bf870010d478ab439c6/YAXLib/Customization/IMemberContext.cs) **PropertyInfo**, **FieldInfo**, and **MemberInfo** properties marked as obsolete, added the [MemberDescriptor ](https://github.com/YAXLib/YAXLib/blob/612f849d5a0ff47c617c0bf870010d478ab439c6/YAXLib/Customization/IMemberContext.cs#L35) property as a replacement for deprecated properties.

**Improvements**:

1. Get rid of Lazy<T> & ThreadLocal<T> as they do impossible using this serializer in Visual Studio Debugger extension [ "The function evaluation requires all threads to run"](https://stackoverflow.com/questions/4460206/lazyt-the-function-evaluation-requires-all-threads-to-run)

2. Add the **DoNotSerializeDefaultValues** option to **YAXSerializationOptions**

3. Add the **TypeInspector** property to **SerializerOptions** (allows type name and members customization without attributes)

4. MemberWrapperCache and UdtWrapperCache: cache the type by type and SerializerOptions.

5. [Fixed failing tests](https://github.com/YAXLib/YAXLib/blob/d5d46550ccc761482e23d51ae3df8936c4e5c558/YAXLibTests/TestHelpers/PoolingHelpers.cs#L38): if choose Get_PooledObject and Parallel_Load_On_Specialized_Pools and run them only, the Parallel_Load_On_Specialized_Pools will fail.

image

### CODE REVIEW

**Comment by axunonb** [comment]
Hi Yevhen

Thanks for the PR. It includes very nice improvements, while it's really huge and will require some time process.
For the moment, YAXLib team has only 1 member left ;-) so you're welcome to join.

Breaking changes
* Yet I'm not aware of all, but the PR includes a couple of breaking changes. Please kindly name the breaking changes introduced in the PR description.
* We had significant breaking changes over last months. So for the sake of other YAXLib users, we should be cautious. Is there a way to avoid or reduce them?

Few things that I came across when going throw the code for the first time:
* SonarCloud Quality Gate fails for the PR. Please have a look at the [Sonar Summary](https://sonarcloud.io/summary/new_code?id=axunonb_YAXLib&branch=pr%2F227) in terms of Maintainability and Coverage.
* Well written XmlDoc is useful for users and maintainers. I was trying hard to get what we currently have, and it should not become less.

Naming conventions
* Clearly, YAXLib does not follow the common naming convention for types. We should not mix `Yax...` and `YAX...` prefixes.
* YAXLib 4.x still has few abbreviations used in member names. `DoNot` is preferred over `Dont`. In the past it was left as is to avoid a breaking change.

What do you think - does this make sense to you?

**Comment by ycherkes** [comment]
Hi @axunonb,
Thanks for the review.

I will try to fix the SonarCloud Quality Gate fails.

I don't clearly understand what you mean when say "breaking changes" - all the functional tests left untouched. The only thing that could be impacted is Default cache sizes. I mean if it was previously cached by **Type** only but now caches by **Type and SerializerOptions**, that means that cache will be filled faster for use cases when serialization/deserialization occurs with multiple SerializerOptions.

**Comment by ycherkes** [comment]
Hi @axunonb,
I know about WellKnownTypes and used it for the [DateTimeOffset serialization](https://github.com/ycherkes/ObjectDumper/blob/experimental-use-yaxlib-instead-of-jsonnet/src/Serialization/Implementation/Xml/DateTimeOffsetKnownType.cs).
But, unfortunately, it doesn't allow [for skipping members at runtime](https://github.com/ycherkes/ObjectDumper/blob/experimental-use-yaxlib-instead-of-jsonnet/src/Serialization/Implementation/XmlSerializer.cs#L114)
In opposition to WellKnownTypes, it provides full control over serialization in runtime, not design time.
It can be a [NamingPolicy](https://github.com/ycherkes/ObjectDumper/blob/experimental-use-yaxlib-instead-of-jsonnet/src/Serialization/Implementation/XmlSerializer.cs#L135) or even [customization of builtin type as DateTime](https://github.com/ycherkes/ObjectDumper/blob/experimental-use-yaxlib-instead-of-jsonnet/src/Serialization/Implementation/Xml/YaxFieldMemberWrapper.cs#L50), which is unavailable by defining KnownType.

**Comment by axunonb** [comment]
Thanks! `WellKnownTypes` can be added/removed at runtime. Skipping members at runtime is the missing feature, I agree, same as the `NamingPolicy` delegate. I like these two, but still would prefer to have these features in the known types instead of `SerializerOptions`, because they are actually type-specific. This should be possible.

As you mentioned built-in types like `DateTime`. You are right, this is an unnecessary limitation. Better move (part of) the logic in `ReflectionUtils.ConvertBasicType(...)` to the known types / known base types? `DateTimeOffset` should be a built-in known type, too, same as with `DateOnly` or `TimeOnly` which I added in v4.1.0.

File: YAXLib/IPropertyInfo.cs [hunk 1]
```diff
@@ -3,11 +3,34 @@
...
-    void SetValue(object? obj, object? value , object[]? index);
+
+    /// <summary>
+    /// Sets the value of the property for the specified object and optional index parameters.
+    /// </summary>
+    /// <param name="obj">The object on which to set the property value.</param>
+    /// <param name="value">The value to be assigned to the property.</param>
+    /// <param name="index">Optional index parameters for indexed properties.</param>
+    void SetValue(object? obj, object? value, object[]? index);
 }
```
Comment (axunonb) [inline YAXLib/IPropertyInfo.cs:36]: 👍

**Comment by ycherkes** [comment]
> Thanks! `WellKnownTypes` can be added/removed at runtime. Skipping members at runtime is the missing feature, I agree, same as the `NamingPolicy` delegate. I like these two, but still would prefer to have these features in the known types instead of `SerializerOptions`, because they are actually type-specific. This should be possible.
> 
> As you mentioned built-in types like `DateTime`. You are right, this is an unnecessary limitation. Better move (part of) the logic in `ReflectionUtils.ConvertBasicType(...)` to the known types / known base types? `DateTimeOffset` and `IpAddress` should be a built-in known type, too, same as with `DateOnly` or `TimeOnly` which I added in v4.1.0.

Agree with you that DateTime as KnownType, the Naming policy implemented in YaxLib sounds like a good plan!
Currently, `ITypeInfoResolver` can help cover features not yet implemented.
Could you rerun the SonarCloud checks?

File: YAXLib/UdtWrapper.cs [hunk 2]
```diff
@@ -406,36 +408,55 @@ internal IEnumerable<MemberWrapper> GetFieldsForDeserialization(bool sorted = tr
...
+
+    private IEnumerable<MemberWrapper> GetMembers()
+    {
+        if (MemberWrapperCache.Instance.TryGetItem((UnderlyingType, _serializerOptions), out var memberWrappers))
+            return memberWrappers;
+
+        IList<IMemberInfo> members = new List<IMemberInfo>();
+
+        foreach (var member in UnderlyingType.GetMembers(
+                     BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public,
```
Comment (axunonb) [inline YAXLib/UdtWrapper.cs:424]: Pls. keep #pragma warning disable S3011 because BindingFlags.NonPublic

Comment (ycherkes) [reply YAXLib/UdtWrapper.cs:424]: Restored #pragma warning disable S3011.

**Comment by axunonb** [comment]
> > Thanks! `WellKnownTypes` can be added/removed at runtime. Skipping members at runtime is the missing feature, I agree, same as the `NamingPolicy` delegate. I like these two, but still would prefer to have these features in the known types instead of `SerializerOptions`, because they are actually type-specific. This should be possible.
> > As you mentioned built-in types like `DateTime`. You are right, this is an unnecessary limitation. Better move (part of) the logic in `ReflectionUtils.ConvertBasicType(...)` to the known types / known base types? `DateTimeOffset` and `IpAddress` should be a built-in known type, too, same as with `DateOnly` or `TimeOnly` which I added in v4.1.0.
> 
> Agree with you that DateTime as KnownType, the Naming policy implemented in YaxLib sounds like a good plan! Currently, `ITypeInfoResolver` can help cover features not yet implemented. Could you rerun the SonarCloud checks?

Would you like to put further changes into this PR?
Another option would be a new branch version/v-next as intermediate step.

**Comment by ycherkes** [comment]
I need some time to refine the code of current PR. No new features. Just code polishing.

**Comment by ycherkes** [comment]
@axunonb Can we add [NetFramework 4.5 support](https://github.com/ycherkes/YAXLib/pull/1/files) to this PR?

**Comment by axunonb** [comment]
Hi Yevhen,
As I can see you're still refactoring the YAXLib codebase. Having a glance at the 2 recent commits, I'd say you're introducing concepts to the better. For me it would be helpful to identify the target outcome of the PR more clearly, also having current users of YAXLib in mind. Maybe we could define a project which includes more steps to reach the goal, eventually publishing new versions on the way. Would be nice if you shared your thoughts.

**Comment by ycherkes** [comment]
Guess, I've already finished this PR. The goal of the two last commits was to make the usage of ITypeInspector easy and do it similarly to [Json.NET.ContractResolver](https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/Serialization/DefaultContractResolver.cs#L61) or [YamlDotNet.TypeInspectorSkeleton](https://github.com/aaubry/YamlDotNet/blob/master/YamlDotNet/Serialization/TypeInspectors/TypeInspectorSkeleton.cs#L30)

**Comment by ycherkes** [comment]
OK. Not a problem. Agree with you.

File: YAXLib/Enums/YAXSerializationOptions.cs [hunk 3]
```diff
@@ -57,5 +57,10 @@ public enum YAXSerializationOptions
...
     /// </summary>
-    StripInvalidXmlChars = 64
-}
\ No newline at end of file
+    StripInvalidXmlChars = 64,
+
+    /// <summary>
+    /// Prevents serialization of default values.
+    /// </summary>
+    DontSerializeDefaultValues = 128
```
Comment (axunonb) [inline YAXLib/Enums/YAXSerializationOptions.cs:65]: Pls. rename as `DoNotSerializeDefaultValues`

File: YAXLib/Customization/IMemberContext.cs [hunk 4]
```diff
@@ -34,11 +22,6 @@ public interface IMemberContext
     /// Gets value of this member in the specified object.
     /// </summary>
     /// <param name="obj">The object from which the value must be retrieved.</param>
-    /// <param name="index">
-    /// Optional index values for indexed properties.
-    /// The indexes of indexed properties are zero-based. This value should be <see langword="null" /> for non-indexed
-    /// properties.
-    /// </param>
     /// <returns>The value for this member.</returns>
-    object? GetValue(object? obj, object[]? index = null);
```
Comment (axunonb) [inline YAXLib/Customization/IMemberContext.cs:43]: Why should the `index` parameter be removed?

File: YAXLib/Customization/IMemberContext.cs [hunk 5]
```diff
@@ -11,19 +9,9 @@ namespace YAXLib.Customization;
 public interface IMemberContext
```
Comment (axunonb) [inline YAXLib/Customization/IMemberContext.cs:9]: `IMemberContext` should be kept compatible with v4.1.0 so there's no need for a new major version

File: YAXLib/Customization/IMemberContext.cs [hunk 6]
```diff
@@ -11,19 +9,9 @@ namespace YAXLib.Customization;
...
 {
     /// <summary>
-    /// The member's <see cref="MemberInfo" /> for member serialization, else <see langword="null" />.
-    /// </summary>
-    MemberInfo MemberInfo { get; }
```
Comment (axunonb) [inline YAXLib/Customization/IMemberContext.cs:16]: Keep, but mark as "Obsolete, will be removed in a future version"

File: YAXLib/Customization/IMemberContext.cs [hunk 7]
```diff
@@ -11,19 +9,9 @@ namespace YAXLib.Customization;
...
-    /// <summary>
-    /// The member's <see cref="FieldInfo" /> for field serialization, else <see langword="null" />.
-    /// </summary>
-    FieldInfo? FieldInfo { get; }
-
-    /// <summary>
-    /// The member's <see cref="PropertyInfo" /> for property serialization, else <see langword="null" />.
+    /// The member's <see cref="IMemberDescriptor" /> for member serialization, else <see langword="null" />.
     /// </summary>
-    PropertyInfo? PropertyInfo { get; }
```
Comment (axunonb) [inline YAXLib/Customization/IMemberContext.cs:26]: Keep, but mark as "Obsolete, will be removed in a future version"

File: YAXLib/Customization/IMemberContext.cs [hunk 8]
```diff
@@ -11,19 +9,9 @@ namespace YAXLib.Customization;
...
+    IMemberDescriptor MemberInfo { get; }
```
Comment (axunonb) [inline YAXLib/Customization/IMemberContext.cs:14]: Maybe rename as `MemberDescription`

File: YAXLib/FieldWrapper.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        return Attribute.GetCustomAttributes(_wrappedFieldInfo);
+    }
+
+    public object? GetValue(object? obj)
+    {
+        return _wrappedFieldInfo.GetValue(obj);
+    }
+
+    public void SetValue(object? obj, object? value)
```
Comment (axunonb) [inline YAXLib/FieldWrapper.cs:37]: How to to set the value of an indexed field?

File: YAXLib/FieldWrapper.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        Type = fieldInfo.FieldType;
+        Name = fieldInfo.Name;
+    }
+
+    public Attribute[] GetCustomAttributes()
+    {
+        return Attribute.GetCustomAttributes(_wrappedFieldInfo);
+    }
+
+    public object? GetValue(object? obj)
```
Comment (axunonb) [inline YAXLib/FieldWrapper.cs:32]: How to to get the value of an indexed field?

File: YAXLib/PropertyWrapper.cs [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+        CanWrite = propertyInfo.CanWrite;
+        IsPublic = ReflectionUtils.IsPublicProperty(propertyInfo);
+    }
+
+    public Attribute[] GetCustomAttributes()
+    {
+        return Attribute.GetCustomAttributes(_wrappedProperty);
+    }
+
+    public object? GetValue(object? obj)
```
Comment (axunonb) [inline YAXLib/PropertyWrapper.cs:33]: See comment for `FieldWrapper`

File: YAXLib/PropertyWrapper.cs [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        return _wrappedProperty.GetValue(obj);
+    }
+
+    public void SetValue(object? obj, object? value)
```
Comment (axunonb) [inline YAXLib/PropertyWrapper.cs:38]: See comment for `FieldWrapper`

**Review by axunonb — CHANGES_REQUESTED** [review]
Make the PR compatible with v4.1.0 to a avoid a new major version.

File: YAXLib/Customization/IMemberContext.cs (hunk 5 above)
Comment (ycherkes) [reply YAXLib/Customization/IMemberContext.cs:11]: Done. Please take a look.

File: YAXLib/Customization/IMemberContext.cs (hunk 6 above)
Comment (ycherkes) [reply YAXLib/Customization/IMemberContext.cs:16]: Done. Please take a look.

File: YAXLib/Customization/IMemberContext.cs (hunk 7 above)
Comment (ycherkes) [reply YAXLib/Customization/IMemberContext.cs:26]: Done. Please take a look.

File: YAXLib/Customization/IMemberContext.cs (hunk 8 above)
Comment (ycherkes) [reply YAXLib/Customization/IMemberContext.cs:17]: Done as MemberDescriptor. Please take a look.

File: YAXLib/FieldWrapper.cs (hunk 9 above)
Comment (ycherkes) [reply YAXLib/FieldWrapper.cs]: Fixed.

File: YAXLib/PropertyWrapper.cs (hunk 11 above)
Comment (ycherkes) [reply YAXLib/PropertyWrapper.cs]: Fixed

File: YAXLib/PropertyWrapper.cs (hunk 12 above)
Comment (ycherkes) [reply YAXLib/PropertyWrapper.cs]: Fixed

File: YAXLib/FieldWrapper.cs (hunk 10 above)
Comment (ycherkes) [reply YAXLib/FieldWrapper.cs]: Fixed.

File: YAXLib/FieldWrapper.cs [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        return WrappedField.GetValue(obj);
+    }
+
+    public void SetValue(object? obj, object? value, object[]? index = null)
+    {
+        WrappedField.SetValue(obj, value);
+    }
+
+    public override string ToString()
```
Comment (axunonb) [inline YAXLib/FieldWrapper.cs:42]: Shouldn't this be `public override string? ToString()`
See https://sonarcloud.io/summary/new_code?id=axunonb_YAXLib&branch=pr%2F227

File: YAXLib/PropertyWrapper.cs [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        return WrappedProperty.GetValue(obj, index);
+    }
+
+    public void SetValue(object? obj, object? value, object[]? index = null)
+    {
+        WrappedProperty.SetValue(obj, value, index);
+    }
+
+    public override string ToString()
```
Comment (axunonb) [inline YAXLib/PropertyWrapper.cs:43]: Shouldn't this be `public override string? ToString()`?
See https://sonarcloud.io/summary/new_code?id=axunonb_YAXLib&branch=pr%2F227
Maybe cover this method with unit test?

**Review by axunonb — CHANGES_REQUESTED** [review]
Excellent, thanks a lot!
We could merge this into a `version/4.2.0` branch,
and then make `DateTime` a known type, and also add `DateTimeOffset` and `IpAddress` as known types as you suggested.
All together would then be the v4.2.0 release. Okay?

**Comment by ycherkes** [comment]
Yup, agree with you. I would also add the KeyValuePair to known types.

**Pull request merged**